core/state/dump.go

```diff
@@ -59,23 +95,44 @@ func (self *StateDB) RawDump() Dump {
 			Nonce:    data.Nonce,
 			Root:     common.Bytes2Hex(data.Root[:]),
 			CodeHash: common.Bytes2Hex(data.CodeHash),
-			Code:     common.Bytes2Hex(obj.Code(self.db)),
-			Storage:  make(map[string]string),
 		}
-		storageIt := trie.NewIterator(obj.getTrie(self.db).NodeIterator(nil))
-		for storageIt.Next() {
-			account.Storage[common.Bytes2Hex(self.trie.GetKey(storageIt.Key))] = common.Bytes2Hex(storageIt.Value)
+		if !excludeCode {
+			account.Code = common.Bytes2Hex(obj.Code(self.db))
+		}
+		if !excludeStorage {
+			account.Storage = make(map[string]string)
+			storageIt := trie.NewIterator(obj.getTrie(self.db).NodeIterator(nil))
+			for storageIt.Next() {
+				account.Storage[common.Bytes2Hex(self.trie.GetKey(storageIt.Key))] = common.Bytes2Hex(storageIt.Value)
+			}
 		}
-		dump.Accounts[common.Bytes2Hex(addr)] = account
+		c.onAccount(common.Bytes2Hex(addr), account)
+	}
+}
+
+// RawDump returns the entire state an a single large object
+func (self *StateDB) RawDump(excludeCode, excludeStorage bool) Dump {
+	dump := &Dump{
+		Accounts: make(map[string]DumpAccount),
 	}
-	return dump
+	self.dump(dump, excludeCode, excludeStorage)
+	return *dump
 }
 
-func (self *StateDB) Dump() []byte {
-	json, err := json.MarshalIndent(self.RawDump(), "", "    ")
+// Dump returns a JSON string representing the entire state as a single json-object
+func (self *StateDB) Dump(excludeCode, excludeStorage bool) []byte {
+	dump := self.RawDump(excludeCode, excludeStorage)
+	json, err := json.MarshalIndent(dump, "", "    ")
```

Comment:
Nitpick, I think we should only use an indent of 2 characters. 4 seems a bit excessive tbh.